server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -0,0 +1,358 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway.remote;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.Version;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.common.Nullable;
+import org.opensearch.common.blobstore.BlobContainer;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Setting.Property;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.gateway.remote.ClusterMetadataMarker.UploadedIndexMetadata;
+import org.opensearch.index.remote.RemoteStoreUtils;
+import org.opensearch.repositories.RepositoriesService;
+import org.opensearch.repositories.Repository;
+import org.opensearch.repositories.blobstore.BlobStoreRepository;
+import org.opensearch.repositories.blobstore.ChecksumBlobStoreFormat;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static org.opensearch.gateway.PersistedClusterStateService.SLOW_WRITE_LOGGING_THRESHOLD;
+
+/**
+ * A Service which provides APIs to upload and download cluster metadata from remote store.
+ *
+ * @opensearch.internal
+ */
+public class RemoteClusterStateService {
+
+    public static final String METADATA_NAME_FORMAT = "%s.dat";
+
+    public static final String METADATA_MARKER_NAME_FORMAT = "%s";
+
+    public static final ChecksumBlobStoreFormat<IndexMetadata> INDEX_METADATA_FORMAT = new ChecksumBlobStoreFormat<>(
+        "index-metadata",
+        METADATA_NAME_FORMAT,
+        IndexMetadata::fromXContent
+    );
+
+    public static final ChecksumBlobStoreFormat<ClusterMetadataMarker> CLUSTER_METADATA_MARKER_FORMAT = new ChecksumBlobStoreFormat<>(
+        "cluster-metadata-marker",
+        METADATA_MARKER_NAME_FORMAT,
+        ClusterMetadataMarker::fromXContent
+    );
+    /**
+     * Used to specify if cluster state metadata should be publish to remote store
+     */
+    public static final Setting<Boolean> REMOTE_CLUSTER_STATE_ENABLED_SETTING = Setting.boolSetting(
+        "cluster.remote_store.state.enabled",
+        false,
+        Property.NodeScope,
+        Property.Final
+    );
```

Comment:
Since this is node level, can divergent configurations cause certain guarantees to break

Reply:
Can you elaborate what do mean by divergent configurations?
If you mean the remote cluster enabled setting is enabled on one node and disabled on the other, then definitely remote cluster state will not work as expected.

Reply:
@Bukhtawar we would perform similar validations as other remote store settings to ensure there is no divergent configuration. following validations will be made
- all nodes joining the cluster have same value for these settings
- if the node attributes also contain remote store setting as true, then remote cluster state setting also should be true